executor/instruction/helpers.rs

```diff
@@ -0,0 +1,172 @@
+                    let checked_tupled = self.filter_map(filter_for_row).map(to_tuple);
+                    TupleIterator::$variant(SortedTupleIterator::new(checked_tupled, tuple_positions, variable_modes))
+                }
+            }
+        )*
+    };
+}
+
+macro_rules! impl_iterator_bound_from {
+    ($($executor:ty : $iter:ty => $variant:ident,)*) => {
+        $(
+            impl ExecutorIteratorBoundFrom<$executor> for $iter {
+                fn bound_from_into_tuple_iterator(self, filter_for_row: Box<FilterMapUnchangedFn<<$executor as BinaryEdgeExecutor>::Element>>, to_tuple: MapToTupleFn<<$executor as BinaryEdgeExecutor>::Element>, tuple_positions: TuplePositions, variable_modes: &VariableModes) -> TupleIterator {
+                    let checked_tupled = self.filter_map(filter_for_row).map(to_tuple);
+                    TupleIterator::$variant(SortedTupleIterator::new(checked_tupled, tuple_positions, variable_modes))
+                }
+            }
+        )*
+    };
+}
+
+impl_iterator_unbound! {
+    HasExecutor: HasIterator => HasSingle,
+    HasReverseExecutor : ChainedHasReverseIterator => HasReverseChained,
+    IsaExecutor: isa_executor::MultipleTypeIsaIterator => IsaUnbounded,
+    IsaReverseExecutor: isa_reverse_executor::MultipleTypeIsaIterator => IsaReverseUnbounded,
+
+    OwnsExecutor: OwnsFlattenedVectorInner => OwnsUnbounded,
+    OwnsReverseExecutor: OwnsFlattenedVectorInner => OwnsReverseUnbounded,
+    RelatesExecutor: RelatesFlattenedVectorInner => RelatesUnbounded,
+    RelatesReverseExecutor: RelatesFlattenedVectorInner => RelatesReverseUnbounded,
+    PlaysExecutor: PlaysFlattenedVectorInner => PlaysUnbounded,
+    PlaysReverseExecutor: PlaysFlattenedVectorInner => PlaysReverseUnbounded,
+    SubExecutor: SubVectorInner => SubUnbounded,
+    SubReverseExecutor: SubVectorInner => SubReverseUnbounded,
+}
+
+impl_iterator_unbound_inverted! {
+    HasExecutor: HasIterator => HasSingle,
+    HasExecutor: KMergeBy<HasIterator, HasOrderingFn> => HasMerged,
+
+    HasReverseExecutor : HasReverseIterator => HasReverseSingle,
+    HasReverseExecutor : KMergeBy<HasReverseIterator, HasOrderingFn> => HasReverseMerged,
+}
+
+impl_iterator_bound_from! {
+    HasExecutor: HasIterator => HasSingle,
+    HasReverseExecutor : HasReverseIterator => HasReverseSingle,
+    IsaExecutor: IsaBoundedInner => IsaBounded,
+    IsaReverseExecutor: isa_reverse_executor::MultipleTypeIsaIterator => IsaReverseBounded,
+    OwnsExecutor: OwnsVectorInner => OwnsBounded,
+    OwnsReverseExecutor: OwnsVectorInner => OwnsReverseBounded,
+    RelatesExecutor: RelatesVectorInner => RelatesBounded,
+    RelatesReverseExecutor: RelatesVectorInner => RelatesReverseBounded,
+    PlaysExecutor: PlaysVectorInner => PlaysBounded,
+    PlaysReverseExecutor: PlaysVectorInner => PlaysReverseBounded,
+    SubExecutor: SubVectorInner => SubBounded,
+    SubReverseExecutor: SubVectorInner => SubReverseBounded,
+}
```

Comment:
I could probably make do with the versions that are used in the TupleIterator enum, rather than the newly introduced `*Inner` type aliases.